### DISCUSSION TITLE
Printing hexadecimals

### DISCUSSION THREAD

**Discussion by cve2022** [micropython/micropython#10728] [Using MicroPython] [answered]
When I use hex() or hexadecimal format with print with the value returned by mem32, if the most significant bit is 1, then I get a signed string.

The only way I could find to print the value as an unsigned int hex was using an array:
```
>>> hex(mem32[0xE000E400])
'-0x7f7f7f80'
>>> hex(array('I', [mem32[0xE000E400]])[0])
'0x80808080'
```
Is there a simpler way?

**Comment by dlech** [comment, marked as answer]
```python
hex(mem32[0xE000E400] & 0xffffffff)
```

**Reply by cve2022** [reply]
Brilliant! Thank you.

PS: for some reason the "Mark as answer" button is not appearing.

**Reply by jimmo** [reply]
@cve2022 That's because this was opened in the "General" category, not one of the Q&A categories. I've moved it to "Using MicroPython".

**Reply by cve2022** [reply]
> @cve2022 That's because this was opened in the "General" category, not one of the Q&A categories. I've moved it to "Using MicroPython".

Thank you for the heads up. 😄

**Comment by cve2022** [comment]
Follow-up question, it will always work? Why? I tested with Python and it works too, FWIW.

I have read here and there (on topics about similar questions) that Python does not have the concept of unsigned int, but this result contradicts that statement. In C, AFAIR, if one has e.g., an int32_t and a uint32_t in the same expression, the int32_t is promoted automatically. Ofc the right way is casting the operands to make sure the result is what one intended. This expression suggests Python behaves the same way. Or am I missing something?

**Reply by bixb922** [reply]
0xffffffff is a positive number that does not fit into 32 bits. Thus the result of the & operation is also stored as an int that exceeds 32 bits.

**Reply by cve2022** [reply]
Oh, I see. -1, for instance, is constituted by gazillions ones (111111111111.......11111111 - binary). When I make a bitwise &, only the least significant 32 bits are preserved, the others are zero. I should have seen that earlier.
